receivers for http/tcp/web socket communication

eleniv3d opened the pull request:
This PR includes:
(1) an http listener that retrieves .ply files from the internet
https://github.com/user-attachments/assets/59ce5c8d-adad-4164-9414-49cebac1f652

(2) a TCP listener that receives pcd color data
https://github.com/user-attachments/assets/a088ad9c-712a-49ea-90fb-4b6deb1e583e

(3) a Websocket listener that receives pcd color data
https://github.com/user-attachments/assets/8310f586-0c51-4769-babe-9453556af1a6

To test the files you can use the senders that are available [here](https://github.com/diffCheckOrg/diffCheck/blob/IO/src/gh/examples/simple_tcp_sender.py) and [here](https://github.com/diffCheckOrg/diffCheck/blob/IO/src/gh/examples/simple_ws_sender.py)

The workflow now is:
- start server in gh (press i_start_btn)
- start sending with vs code
- load pcd in gh (press i_load_btn)

I added automatic btns and panels for standard inputs like we had for the visualization_settings component and moved these utilities to a new df_gh_canvas.py file so that they are available to use in components in the future. The components look like this when dropped in the carvas :

<img width="1042" alt="automatic sliders" src="https://github.com/user-attachments/assets/97a8bdbf-b34d-47dc-a828-219de595e595" />

eleniv3d committed:
ADD Http Listener

eleniv3d committed:
ADD TCP Listener

9and3 commented:
> @DamienGilliard , @9and3 I run pre-commit but linter doesn't like the ghenv.Component.Message without importing ghenv but I think it only works like this :)

You can flag the line with an in-line comment
```py
param = ghenv.Component.Params.Input[indx]  # noqa: F821
```
see https://github.com/diffCheckOrg/diffCheck/blob/a73bc9dca0bc8a8dffab7214dac0b5084a4f2a58/src/gh/components/DF_csv_exporter/code.py#L30-L42

The linter should stop complaining about it :)

eleniv3d committed:
FIX lintering

eleniv3d commented:
> > @DamienGilliard , @9and3 I run pre-commit but linter doesn't like the ghenv.Component.Message without importing ghenv but I think it only works like this :)
> 
> You can flag the line with an in-line comment
> 
> ```python
> param = ghenv.Component.Params.Input[indx]  # noqa: F821
> ```
> 
> see
> 
> https://github.com/diffCheckOrg/diffCheck/blob/a73bc9dca0bc8a8dffab7214dac0b5084a4f2a58/src/gh/components/DF_csv_exporter/code.py#L30-L42
> 
> The linter should stop complaining about it :)

Thanks @9and3! it seems to have done the trick ;)

eleniv3d committed:
FIX tcp component + add example file sender

eleniv3d committed:
Merge pull request #148 from diffCheckOrg/patch_ws

Patch: CI + gh fixes

eleniv3d committed:
Merge branch 'main' into IO

eleniv3d committed:
WIP FIX WS Receiver

eleniv3d committed:
FIX websocket component

eleniv3d commented:
All three are working now!

Remains to be verified if I successfully added websockets to the python package or it just works on my side because I had it already installed. :)

eleniv3d committed:
FIX light refactoring and input reorder

eleniv3d committed:
ADD drop btns and panel with standard inputs similar to vizualization settings

DamienGilliard commented on src/gh/diffCheck/diffCheck/df_gh_canvas.py:
I don't know if I'm being picky, but it seems to me that calling the file df_gh_canvas_utils.py would be more explicit. Otherwise thanks for this new module !

eleniv3d replied:
Renamed! :)

DamienGilliard commented on src/gh/components/DF_tcp_listener/code.py:
```diff
@@ -0,0 +1,136 @@
+            df_gh_canvas.add_button(
+                ghenv.Component, label, idx, x_offset=60)  # noqa: F821
+        df_gh_canvas.add_panel(ghenv.Component, "Host", "127.0.0.1", 3, 60, 20)  # noqa: F821
+        df_gh_canvas.add_panel(ghenv.Component, "Port", "5000", 4, 60, 20)  # noqa: F821
```
clean ;)

eleniv3d replied:
it is clean or it should be cleaned? :P

DamienGilliard replied:
I meant it is clean ;)

DamienGilliard commented on src/gh/components/DF_tcp_listener/code.py:
```diff
@@ -0,0 +1,136 @@
+        sc.sticky.setdefault(f'{prefix}_prev_load', False)
+
+        # Client handler
+        def handle_client(conn):
```
Could you add doctrings (here and elsewhere) in your functions that describe the types of data needed ? It is not clear to me what "conn" is :( . For example:
```py
def handle_client(conn: <type_of_conn>):
    """
    reads the incoming bytes from a single client socket and stores valid data in a shared buffer
    :param conn: <what conn is in pain english>
    :returns: <what it returns>
    """
```
Like what you did for some of the functions in the new `df_gh_canvas.py` module.
I noticed that I had not done it myself so I'll fix it and put full docstring everywhere too ;)

eleniv3d replied:
Done ✅

DamienGilliard commented on src/gh/components/DF_tcp_listener/code.py:
```diff
@@ -0,0 +1,136 @@
+            reads the incoming bytes from a single client socket and stores valid data in a shared buffer
+            """
+            buf = b''
+            with conn:
+                while sc.sticky.get(f'{prefix}_server_started', False):
+                    try:
+                        chunk = conn.recv(4096)
+                        if not chunk:
+                            break
+                        buf += chunk
+                        while b'\n' in buf:
+                            line, buf = buf.split(b'\n', 1)
+                            try:
+                                raw = json.loads(line.decode())
+                            except Exception:
+                                continue
+                            if isinstance(raw, list) and all(isinstance(pt, list) and len(pt) == 6 for pt in raw):
+                                sc.sticky[f'{prefix}_cloud_buffer_raw'] = raw
+                    except Exception:
+                        break
```
In my experience, having infinite loops in separate threads that run "full gas" can lead to issues with the cpu, I would add a small `time.sleep(0.1)` for 100ms sleep before re-running the loop, for example.
Although here on my computer this loop without pause seems to have no negative impact

eleniv3d replied:
thanks for the comment!  I added a time.sleep 🕰️

DamienGilliard commented on src/gh/components/DF_tcp_listener/code.py:
```diff
@@ -0,0 +1,136 @@
+            sc.sticky[f'{prefix}_status_message'] = 'Stopped'
+
+        # Start or stop server based on inputs
+        if i_start and not sc.sticky[f'{prefix}_prev_start']:
+            start_server()
```
When testing your components I made the mistake to first try to send data to the TCP socket (before starting the server, dumb of me) and it lead me to think that maybe we could add after this `if` something like:
```py
else:
    self.AddRuntimeMessage(RML.Warning, "Please start server here before trying to send data from remote device.")
```
Maybe it is because I'm dumb but I think it could improve slightly the user-friendliness

eleniv3d replied:
lol :p it makes a lot of sense. i added it!

DamienGilliard commented on src/gh/components/DF_websocket_listener/code.py:
```diff
@@ -0,0 +1,133 @@
+            cnt = len(sc.sticky[f'{prefix}_loaded_pcd']) if sc.sticky[f'{prefix}_loaded_pcd'] else 0
+            logs.append(f"Loaded pcd with {cnt} pts")
+            ghenv.Component.ExpireSolution(True)  # noqa: F821
+
```
Here too, as for the TCP listener, I would add a small warning to start server in gh before trying to send data. It sounds obvious but I think it's a nice-to-have feature. Let  us know what you think ;)

eleniv3d replied:
Done!

DamienGilliard commented on src/gh/diffCheck/diffCheck.egg-info/PKG-INFO:
```diff
@@ -11,14 +11,7 @@ Classifier: Programming Language :: Python :: 3.9
 Description-Content-Type: text/markdown
 Requires-Dist: numpy
 Requires-Dist: pybind11>=2.5.0
-Dynamic: author
-Dynamic: author-email
-Dynamic: classifier
-Dynamic: description
-Dynamic: description-content-type
-Dynamic: home-page
-Dynamic: requires-dist
-Dynamic: summary
```
Why have these been deleted? is it intentional ?

eleniv3d replied:
I guess this file auto-generated?

eleniv3d replied:
Anyhow I put them back! :)

DamienGilliard commented on src/gh/components/DF_websocket_listener/code.py:
```diff
@@ -0,0 +1,133 @@
+            i_host: str,
+            i_port: int):
+
+        prefix = 'ws'
+
+        # Persistent state across runs
+        sc.sticky.setdefault(f'{prefix}_ws_server', None)
+        sc.sticky.setdefault(f'{prefix}_ws_loop',   None)
+        sc.sticky.setdefault(f'{prefix}_ws_thread', None)
+        sc.sticky.setdefault(f'{prefix}_last_pcd',  None)
+        sc.sticky.setdefault(f'{prefix}_loaded_pcd', None)
+        sc.sticky.setdefault(f'{prefix}_ws_logs',   [])
+        sc.sticky.setdefault(f'{prefix}_ws_thread_started', False)
+        sc.sticky.setdefault(f'{prefix}_prev_start', False)
+        sc.sticky.setdefault(f'{prefix}_prev_stop',  False)
+        sc.sticky.setdefault(f'{prefix}_prev_load',  False)
```
having `ws_ws_server` and other ws_ws names seems strange to me. Is it necessary ?

eleniv3d replied:
by accident! fixed!

DamienGilliard reviewed (commented):
Small questions here and there, but overall thanks a lot for this great addition!

eleniv3d committed:
FIX rename python file

eleniv3d committed:
FIX remove repeated ws from stickys

eleniv3d committed:
ADD warning if trying to load data from tcp/ws listener without starting the server first

eleniv3d committed:
ADD sleep to prevent CPU spin

eleniv3d committed:
ADD doc strings

eleniv3d committed:
FIX add docstrings

eleniv3d committed:
FIX missing lines from PKG-INFO

eleniv3d committed:
FIX RML.Warning

eleniv3d commented:
@DamienGilliard 

I think i implemented all the comments and didnt break anything :)

One thing that is not working is the warning for starting the servers first. Not sure why :/
Alternatively, I added a component message that says to "start the server first!" if you press load without the server running

DamienGilliard commented:
This is good to merge I believe ?

eleniv3d committed:
FIX cleanup warnings that didnt work

eleniv3d commented:
@DamienGilliard yes! i took out the warnings that didn't work like we discussed Friday. Feel free to merge it :)
